pandas/core/arrays/datetimelike.py

```diff
@@ -305,6 +305,12 @@ def asi8(self):
     def nbytes(self):
         return self._data.nbytes
 
+    def __array__(self, dtype=None):
+        # used for Timedelta/DatetimeArray, overwritten by PeriodArray
```

Comment:
i doubt we test this, but why not astype here if ``dtype`` is not (object, None), then return ``__array__`` on that? which possibly would recurse but should be ok

Reply:
Hmm in my quick experiments, that gets into a recursion error pretty quickly. We would need to be careful to not use __array__ in anything that `.astype` calls

1. `_format_native_types`
2. `Categorical.__init__`
3. `PeriodIndex.asfreq`

and currently the fallback on `astype` for unrecognized dtypes is to call `np.asarray(self, dtype=dtype)`, so that would need to be changed as well.